From 42dd726100a26b26fb4040e70bf2b4cbe59dbfc0 Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Tue, 18 Feb 2020 17:38:32 +0100
Subject: [PATCH 26/26] Debug _PyImport_FindSharedFuncptrWindows() values of
 _PyInterpreterState_Get()

Hidden behind env var CONDA_DEBUG_INTERP_STATE
---
 Python/dynload_win.c | 63 ++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 63 insertions(+)

diff --git a/Python/dynload_win.c b/Python/dynload_win.c
index 56d8913455..4091efcd51 100644
--- a/Python/dynload_win.c
+++ b/Python/dynload_win.c
@@ -190,6 +190,8 @@ dl_funcptr _PyImport_FindSharedFuncptrWindows(const char *prefix,
 #if HAVE_SXS
         ULONG_PTR cookie = 0;
 #endif
+        int debug_it = _wgetenv(L"CONDA_DEBUG_INTERP_STATE") ? 1 : 0;
+        PyInterpreterState* ts = 0;
 
         /* Don't display a message box when Python can't load a DLL */
         old_mode = SetErrorMode(SEM_FAILCRITICALERRORS);
@@ -202,13 +204,74 @@ dl_funcptr _PyImport_FindSharedFuncptrWindows(const char *prefix,
            AddDllDirectory function. We add SEARCH_DLL_LOAD_DIR to
            ensure DLLs adjacent to the PYD are preferred. */
         /* This resyncs values in PATH to AddDllDirectory() */
+
+
+        if (debug_it) {
+            ts = _PyInterpreterState_Get();
+            if (!ts)
+            {
+                fwprintf(stderr, L"(0) ts = 0\n");
+            }
+            else
+            {
+                fwprintf(stderr, L"(0) ts is OK\n");
+            }
+        }
+
         extern int CondaEcosystemModifyDllSearchPath(int, int);
         CondaEcosystemModifyDllSearchPath(1, 1);
+        if (debug_it) {
+            ts = _PyInterpreterState_Get();
+            if (!ts)
+            {
+                fwprintf(stderr, L"(1) ts = 0\n");
+            }
+            else
+            {
+                fwprintf(stderr, L"(1) ts is OK\n");
+            }
+        }
 
         Py_BEGIN_ALLOW_THREADS
+
+        if (debug_it) {
+/* - Will always crash            ts = _PyInterpreterState_Get();
+            if (!ts)
+            {
+                fwprintf(stderr, L"(2) ts = 0\n");
+            }
+            else
+            {
+                fwprintf(stderr, L"(2) ts is OK\n");
+            } */
+        }
         hDLL = LoadLibraryExW(wpathname, NULL,
                               LOAD_WITH_ALTERED_SEARCH_PATH);
+        if (debug_it) {
+/* - Will always crash            ts = _PyInterpreterState_Get();
+            if (!ts)
+            {
+                fwprintf(stderr, L"(3) ts = 0\n");
+            }
+            else
+            {
+                fwprintf(stderr, L"(3) ts is OK\n");
+            } */
+        }
+
         Py_END_ALLOW_THREADS
+
+        if (debug_it) {
+            ts = _PyInterpreterState_Get();
+            if (!ts)
+            {
+                fwprintf(stderr, L"(4) ts = 0\n");
+            }
+            else
+            {
+                fwprintf(stderr, L"(4) ts is OK\n");
+            }
+        }
 #if HAVE_SXS
         _Py_DeactivateActCtx(cookie);
 #endif
-- 
2.25.0

